frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,842 @@
+	pub fn get() -> Self {
+		SignedSubmissions {
+			indices: SignedSubmissionIndices::<T>::get(),
+			next_idx: SignedSubmissionNextIndex::<T>::get(),
+			insertion_overlay: BTreeMap::new(),
+			deletion_overlay: BTreeSet::new(),
+		}
+	}
+
+	/// Put the signed submissions back into storage.
+	pub fn put(mut self) {
+		SignedSubmissionIndices::<T>::put(self.indices);
+		SignedSubmissionNextIndex::<T>::put(self.next_idx);
+		for key in self.deletion_overlay {
+			self.insertion_overlay.remove(&key);
+			SignedSubmissionsMap::<T>::remove(key);
+		}
+		for (key, value) in self.insertion_overlay {
+			SignedSubmissionsMap::<T>::insert(key, value);
+		}
+	}
+
+	/// Get the submission at a particular index.
+	fn get_submission(&self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		if self.deletion_overlay.contains(&idx) {
+			// Note: can't actually remove the item from the insertion overlay (if present)
+			// because we don't want to use `&mut self` here. There may be some kind of
+			// `RefCell` optimization possible here in the future.
+			None
+		} else {
+			self.insertion_overlay
+				.get(&idx)
+				.cloned()
+				.or_else(|| SignedSubmissionsMap::<T>::try_get(idx).ok())
+		}
+	}
+
+	/// Perform three operations:
+	///
+	/// - Remove a submission (identified by score)
+	/// - Insert a new submission (identified by score and insertion index)
+	/// - Return the weakest submission
+	///
+	/// Note: in the case that `weakest_score` is not present in `self.indices`, this will return
+	/// `None` without inserting the new submission and without further notice.
+	///
+	/// Note: this does not enforce any ordering relation between the submission removed and that
+	/// inserted.
+	fn swap_out_submission(
+		&mut self,
+		remove_score: ElectionScore,
+		insert: Option<(ElectionScore, u32)>,
+	) -> Option<SignedSubmissionOf<T>> {
+		let remove_idx = self.indices.remove(&remove_score)?;
+		if let Some((insert_score, insert_idx)) = insert {
+			self.indices
+				.try_insert(insert_score, insert_idx)
+				.expect("just removed an item, we must be under capacity; qed");
+		}
+
+		self.insertion_overlay.remove(&remove_idx).or_else(|| {
+			(!self.deletion_overlay.contains(&remove_idx)).then(|| {
+				self.deletion_overlay.insert(remove_idx);
+				SignedSubmissionsMap::<T>::try_get(remove_idx).ok()
+			}).flatten()
+		})
+	}
+
+	/// Iterate through the set of signed submissions in order of increasing score.
+	pub fn iter(&self) -> impl '_ + Iterator<Item = SignedSubmissionOf<T>> {
+		self.indices.iter().map(move |(_score, idx)| {
+			self.get_submission(*idx).expect("SignedSubmissions must remain internally consistent")
+		})
+	}
+
+	/// Empty the set of signed submissions, returning an iterator of signed submissions in
+	/// arbitrary order.
+	///
+	/// Note that if the iterator is dropped without consuming all elements, not all may be removed
+	/// from the underlying `SignedSubmissionsMap`, putting the struct into an invalid state.
```

Comment:
```suggestion
	/// from the underlying `SignedSubmissionsMap`, putting the storages into an invalid state.
```
I think what is invalid in this case is `SignedSubmissionsIndices` storage.